Core hole calculations with MOM / IMOM

msnik1999 opened the pull request:
Adding functionality for modifying the orbital occupancies to allow core hole calculations, e.g. for XPS core binding energy calculations. @lratcliff
Also adding MOM/IMOM methods to assist these calculations.

lratcliff committed:
initial conversion of occupancies from integer to double

lratcliff committed:
Initial implementation of core hole calculations

lratcliff committed:
Switching logic so that restricted checks depend on doubly occ (only happens for rest) not singly occ (which could be unrest or rest with a hole)

lratcliff committed:
XX with occupancies

lratcliff committed:
updating input file for explicit occupancies

lratcliff committed:
fixing occupancies/spin factors for XX in restricted calculations

lratcliff committed:
removing unneeded print statement

lratcliff committed:
adding positions of the orbitals to enable association with a particular atomic orbital

Commit:
first try

Commit:
first try

Commit:
restricted MOM

Commit:
first try for IMOM; only IMOM in this version

Commit:
first try for IMOM; only IMOM in this version

Commit:
first cleanup, code now readable, but longer

Commit:
first cleanup, code now readable, but longer

Commit:
small bugfix

Commit:
added input in input parser for MOM/IMOM

Commit:
added input in input parser for MOM/IMOM

Commit:
small bugfix

msnik1999 committed:
new logic to treat different occupation numbers; cleanup

msnik1999 committed:
new logic to treat different occupation numbers; cleanup

msnik1999 committed:
small formatting

msnik1999 committed:
small formatting

msnik1999 committed:
small temporary fix in restricted occupations

msnik1999 committed:
print orbital occupations as doubles instead of integers

msnik1999 committed:
for now only removal of one electron possible#

msnik1999 committed:
small bugfix in ExchangePotentialD2.cpp

msnik1999 committed:
bugfix in initial guess

msnik1999 committed:
IMOM: fetch Orbitals before calculation of initial guess energy; some general cleanup; added first test for DeltaSCF

msnik1999 committed:
IMOM: fetch Orbitals before calculation of initial guess energy; some general cleanup; added first test for DeltaSCF

msnik1999 committed:
small bugfix in test

msnik1999 committed:
small bugfix in test

msnik1999 committed:
small bugfix in driver.cpp and some changes to the core-hole test

msnik1999 committed:
small bugfix in driver.cpp and some changes to the core-hole test

msnik1999 committed:
added tests

msnik1999 committed:
added tests

msnik1999 committed:
changing copies to deep_copies

msnik1999 committed:
changing copies to deep_copies

msnik1999 committed:
some cleanup + some more output depending on print level

msnik1999 committed:
some cleanup + some more output depending on print level

msnik1999 committed:
- Implemented partial localization for different occupations
- Fixed values for tests
- some cleanup

msnik1999 committed:
- Implemented partial localization for different occupations
- Fixed values for tests
- some cleanup

msnik1999 committed:
small bugfix for unrestricted MOM/IMOM calculations

msnik1999 committed:
small bugfix for unrestricted MOM/IMOM calculations

ilfreddy commented:
1. To make the test pass, you should first rebase so that you make sure your code also is compiling&testing the macos-12 build
2. We should see what is missing to bring this from draft to mergeable PR
3. Does this PR depend on the the corresponding one (fractional occupation numbers) in mrcpp?

msnik1999 committed:
remove partial localization

msnik1999 committed:
remove partial localization

msnik1999 committed:
added tests for unrestricted IMOM; some cleanup

msnik1999 committed:
added tests for unrestricted IMOM; some cleanup

msnik1999 committed:
small cleanup

msnik1999 committed:
small cleanup

msnik1999 commented:
> 1. To make the test pass, you should first rebase so that you make sure your code also is compiling&testing the macos-12 build
> 
>     2. We should see what is missing to bring this from draft to mergeable PR
> 
>     3. Does this PR depend on the the corresponding one (fractional occupation numbers) in mrcpp?

I think I addressed all issues and problems. This PR is dependent on the corresponding one in MRCPP.

stigrj commented:
The MRCPP PR is now merged, so you can update the `GIT_TAG` in `external/upstream/fetch_mrcpp.cmake` to the latest on `master`. Also, for the failing exchange issue, you can add an explicit exception somewhere in the initialization that fails whenever someone is trying to run MOM calculations with exact exchange. Then open an issue that explains the problem.

msnik1999 committed:
incorporated PR comments

msnik1999 commented:
> The MRCPP PR is now merged, so you can update the `GIT_TAG` in `external/upstream/fetch_mrcpp.cmake` to the latest on `master`. Also, for the failing exchange issue, you can add an explicit exception somewhere in the initialization that fails whenever someone is trying to run MOM calculations with exact exchange. Then open an issue that explains the problem.

Done

Gabrielgerez commented:
Ill try resolving the conflict and we can probably just merge it after that

msnik1999 committed:
make code mostly compilable

msnik1999 commented:
ToDo:

- [ ] Cleanup
- [ ] Check documentation
- [ ] Changelog?
- [ ] Check/fix compilation issues
- [ ] Test all features again, as code was written more than 1 year ago
- [ ] Fix Exchange issues? (Issue #493)

msnik1999 committed:
cleanup

msnik1999 committed:
fixed bug in calculation of number of electrons that lead to failing tests

msnik1999 committed:
updated test references for core-hole and IMOM tests

msnik1999 committed:
fixed MPI print issues, split tests to avoid timeout

msnik1999 committed:
added missing files

msnik1999 committed:
updated tests

msnik1999 commented:
The tests should now pass, from my side this PR is ready to be merged. @ilfreddy

ilfreddy reviewed (commented):
There is one point I would like to revise. @msnik1999, we can discuss it tomorrow.

msnik1999 committed:
storing mom orbitals in the Molecule object
